#  Differential Expression Analysis for Ulcerative colitis and Crohn's disease

The dataset used in this project has been taken from GEO Series:GSE235236

**[Data here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE235236)**

File : GSE235236_RAW.tar was simplified into a csv using ChatGPT Model 4o. The code for it is not yet available

## First look at the dataset

In [2]:
import pandas as pd

def first_look(data_path):
    csv_data = pd.read_csv(data_path)
    print("The shape of the data is ")
    display(csv_data.shape)
    print("The first five columns of the dataset are")
    display(csv_data.head())
    return csv_data

csv_data = first_look("ibd_cd_gene_expression.csv")




The shape of the data is 


(58278, 49)

The first five columns of the dataset are


,gene_id,CD_1,UC_1,UC_2,CD_2,UC_3,CD_3,CD_4,CD_5,CD_6,...,UC_22,CD_18,CD_19,UC_23,UC_24,UC_25,CD_20,CD_21,UC_26,CD_22
0,ENSG00000000003.14_TSPAN6,1664.00,614.00,2507.0,2459.00,1604.00,1983.00,2067.00,530.00,1176.00,...,1231.00,651.00,2425.0,1164.00,670.00,2459.00,2494.00,1062.00,2205.00,2242.00
1,ENSG00000000005.5_TNMD,4.00,3.00,12.0,11.00,15.00,13.00,12.00,1.00,9.00,...,8.00,6.00,5.0,17.00,12.00,11.00,15.00,5.00,8.00,23.00
2,ENSG00000000419.12_DPM1,677.00,675.00,701.0,644.00,903.00,655.00,827.00,1368.00,689.00,...,697.00,572.00,723.0,709.00,700.00,614.00,570.00,778.00,776.00,649.00
3,ENSG00000000457.13_SCYL3,392.01,379.55,511.8,433.14,342.08,380.24,387.16,260.52,377.84,...,419.19,325.56,397.3,457.94,309.15,437.38,417.88,384.70,414.99,347.62
4,ENSG00000000460.16_C1orf112,111.17,169.45,105.2,122.86,111.92,76.37,189.11,229.48,96.16,...,230.94,192.81,136.7,207.19,148.85,144.62,130.12,159.52,182.01,149.38


Here in this dataset we can see there are 58,000 Genes and 48 different sample of people.
lets try to count the number of people with UC and Crohn's

In [3]:

def clean_column_names(data):
        column_name = data.columns
        cleaned_columns = column_name.str.replace(r"[^a-zA-Z]","",regex=True)
        data.columns = cleaned_columns
        return data

csv_data = clean_column_names(csv_data)
csv_data.columns[1:].value_counts()



UC    26
CD    22
Name: count, dtype: int64

Here we have 26 sample for UC and 22 Samples for Crohn's which we will use for our diffrential Gene Analysis

## Count per Million Normalization

The gene counts that we see here are not directly comparable as the count of the gene also depends on sequencing depth (Total number of reads collected). So here I'll be performing CPM normalization. By dividing by total reads we can make the samples comparable 

In [4]:
def cpm_normalization(data):
    data.iloc[:,1:]= data.iloc[:,1:].div(data.iloc[:,1:].sum(axis=0),axis=1) * 1_000_000
    return data

csv_data = cpm_normalization(csv_data)
display(csv_data)


,geneid,CD,UC,UC,CD,UC,CD,CD,CD,CD,...,UC,CD,CD,UC,UC,UC,CD,CD,UC,CD
0,ENSG00000000003.14_TSPAN6,69.298029,27.159271,98.871895,98.982732,71.174978,81.628948,78.961016,20.567514,53.066669,...,54.827707,30.937702,99.509017,50.457724,32.108108,98.672869,108.916442,46.397939,94.690058,91.964777
1,ENSG00000000005.5_TNMD,0.166582,0.132700,0.473260,0.442786,0.665601,0.535137,0.458409,0.038807,0.406122,...,0.356313,0.285140,0.205173,0.736926,0.575071,0.441400,0.655071,0.218446,0.343547,0.943439
2,ENSG00000000419.12_DPM1,28.193970,29.857505,27.646270,25.923090,40.069205,26.962663,31.592047,53.087471,31.090931,...,31.043795,27.183357,29.668049,30.734129,33.545784,24.638122,24.892691,33.990204,33.324029,26.621383
3,ENSG00000000457.13_SCYL3,16.325433,16.788764,20.184538,17.435291,15.179262,15.652340,14.789815,10.109903,17.049924,...,18.670371,15.471702,16.303065,19.851039,14.815256,17.550850,18.249400,16.807238,17.821055,14.259053
4,ENSG00000000460.16_C1orf112,4.629725,7.495339,4.148912,4.945514,4.966274,3.143723,7.224150,8.905346,4.339193,...,10.285874,9.162977,5.609436,8.981388,7.133271,5.803201,5.682521,6.969302,7.816117,6.127430
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
58273,ENSG00000284592.1_RP11-71G12.2,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
58274,ENSG00000284594.1_MIR7847,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
58275,ENSG00000284595.1_MIR6785,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.026203,0.000000,0.000000,0.000000
58276,ENSG00000284596.1_MIR4467,0.000000,0.000000,0.037861,0.000000,0.000000,0.000000,0.036673,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000


In [5]:
sums_series = csv_data.iloc[:,1:].sum(axis=0)
display(sums_series)

CD    1000000.0
UC    1000000.0
UC    1000000.0
CD    1000000.0
UC    1000000.0
CD    1000000.0
CD    1000000.0
CD    1000000.0
CD    1000000.0
UC    1000000.0
CD    1000000.0
CD    1000000.0
UC    1000000.0
CD    1000000.0
CD    1000000.0
UC    1000000.0
UC    1000000.0
CD    1000000.0
UC    1000000.0
UC    1000000.0
UC    1000000.0
UC    1000000.0
UC    1000000.0
UC    1000000.0
CD    1000000.0
CD    1000000.0
UC    1000000.0
CD    1000000.0
UC    1000000.0
UC    1000000.0
UC    1000000.0
UC    1000000.0
UC    1000000.0
CD    1000000.0
CD    1000000.0
CD    1000000.0
UC    1000000.0
UC    1000000.0
UC    1000000.0
CD    1000000.0
CD    1000000.0
UC    1000000.0
UC    1000000.0
UC    1000000.0
CD    1000000.0
CD    1000000.0
UC    1000000.0
CD    1000000.0
dtype: float64

Here we can see all samples sum to 1,000,000, this means they have been successfully normalized

Now I'll remove which have very low expression, here I'll set the threshold to 10, and remove any gene that have expression less than 10.

In [6]:

def apply_cpm_threshold(data,col_name1,col_name2,threshold):

    def select_column(data,column):
        column_bool_list = data.columns == column
        column_data = csv_data.loc[:,column_bool_list]
        return column_data

    column1_data = select_column(data,col_name1)
    column2_data = select_column(data,col_name2)

    keep_gene = (column1_data.mean(axis=1) > threshold) | (column2_data.mean(axis=1)>threshold)
    data = data.loc[keep_gene,:]
    
    return data

csv_data = apply_cpm_threshold(csv_data,"UC","CD",10)
csv_data



,geneid,CD,UC,UC,CD,UC,CD,CD,CD,CD,...,UC,CD,CD,UC,UC,UC,CD,CD,UC,CD
0,ENSG00000000003.14_TSPAN6,69.298029,27.159271,98.871895,98.982732,71.174978,81.628948,78.961016,20.567514,53.066669,...,54.827707,30.937702,99.509017,50.457724,32.108108,98.672869,108.916442,46.397939,94.690058,91.964777
2,ENSG00000000419.12_DPM1,28.193970,29.857505,27.646270,25.923090,40.069205,26.962663,31.592047,53.087471,31.090931,...,31.043795,27.183357,29.668049,30.734129,33.545784,24.638122,24.892691,33.990204,33.324029,26.621383
3,ENSG00000000457.13_SCYL3,16.325433,16.788764,20.184538,17.435291,15.179262,15.652340,14.789815,10.109903,17.049924,...,18.670371,15.471702,16.303065,19.851039,14.815256,17.550850,18.249400,16.807238,17.821055,14.259053
5,ENSG00000000938.12_FGR,0.832909,27.203504,3.076190,10.143818,32.570096,12.143489,12.950065,92.903074,11.145805,...,23.605755,19.009341,6.811751,5.418570,27.459620,6.741375,4.323467,48.320264,3.864900,10.213751
6,ENSG00000000971.15_CFH,9.499327,72.211370,29.853239,27.538856,14.126281,17.997475,36.241464,29.923017,52.313086,...,45.947044,51.385099,15.531203,27.730074,70.715951,48.676743,9.170991,16.587045,48.289782,14.287356
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
57675,ENSG00000283496.1_ZNF511-PRAP1,17.560637,3.265305,18.905553,5.073922,29.258063,15.583184,9.332833,4.890412,34.352094,...,7.927971,12.699665,10.776929,20.650820,16.300376,8.035077,14.863120,17.460829,5.291907,18.675577
57867,ENSG00000283782.1_AC116366.7,9.456432,16.634390,3.771093,22.626350,7.511090,17.752958,20.072219,13.672352,5.377964,...,12.189478,9.943786,8.128143,20.946891,17.249242,4.404365,11.345390,16.221366,7.661091,12.937008
57928,ENSG00000283907.1_LLNLF-96A1.1,92.286318,111.025684,111.531599,60.218856,322.727939,128.474002,103.753324,1.862718,177.701140,...,131.835916,229.252650,280.020427,234.949194,97.953690,187.193547,86.862391,447.290110,61.237198,79.002748
57940,ENSG00000283932.1_RP4-788L20.5,14.702093,10.088740,10.846724,36.250866,19.969374,18.542491,47.852974,29.194617,16.905073,...,20.889312,9.482810,48.703611,32.539163,12.917044,45.059674,39.547499,20.600772,63.762267,24.368205


In [7]:
csv_data.shape

(10445, 49)

After filtering genes based on CPM threshold we are left with around 10,445 genes. We started with 58,278. This makes further analysis much easier

## Diffrential Gene Analysis

### Kolmogorov-Smirnov Test (KS)


KS test is not commonly used in differential gene expression analysis because it does not specifically target mean differences, which are usually of biological interest. However, it is useful in a beginner setting because it is simple to implement and does not require strong assumptions.

In this project, KS test compares the distribution of gene expression values across samples (26 UC vs 22 CD) and identifies genes whose expression patterns differ between conditions.

It can detect differences beyond just the mean, such as variability, skewness, or subpopulation effects. However, for typical differential expression tasks where consistent up/down regulation is the goal, methods like DESeq2 are more appropriate and statistically powerful.

Now our first step is to divide our samples into 2 different data frames. the first data frame would be for UC and the other CD for applying the ks test

In [ ]:
def get_only_column_with_class(data,column_name):
    column_df = data.loc[:,[column_name]]
    return column_df
uc_df = get_only_column_with_class(csv_data,"UC")
cd_df = get_only_column_with_class(csv_data,"CD")
display(uc_df.shape)
display(cd_df.shape)

(10445, 26)

(10445, 22)

In [ ]:
from scipy.stats import ks_2samp
import numpy as np

def ks_test_to_dataframe(df1,df2):
    p_values = []
    ks_stats = []

    for i in range(df1.shape[0]):
        
        df1_values = df1.iloc[i, :].values
        df2_values = df2.iloc[i, :].values

        stat, p = ks_2samp(df1_values, df2_values)
        
        ks_stats.append(stat)
        p_values.append(p)

    p_values = np.array(p_values)
    ks_stats = np.array(ks_stats)

    return p_values,ks_stats

p_values,ks_stats = ks_test_to_dataframe(uc_df,cd_df)



array([0.38206672, 0.91757654, 0.25164218, ..., 0.88597959, 0.88597959,
       0.77340635], shape=(10445,))

### Implimenting log2fold

log2fold tells us the differenece between gene expression across the condition. 

It exactly tells us how expression changes between conditions

- +1 means 2 times more expressed in one condtion 
- +2 means 4 times more expressed in one condtion 
- -1 means 2 times less expressed in one condtion 
- -2 means 4 times less expressed in one condtion 

This gives a very good intuition to judge the genes 

In [36]:
def log2fold(df1,df2):

    avoid_divide_by_zero = 1 
    log2_fold_change_list = []

    for i in range(df1.shape[0]):

        df1_value = df1.iloc[i,:].values
        df2_value = df2.iloc[i,:].values
    
        df1_mean = df1_value.mean()
        df2_mean = df2_value.mean()
    
        fold_change = (df1_mean + avoid_divide_by_zero)/(df2_mean + avoid_divide_by_zero)
        log2_fold_change = np.log2(fold_change)
    
        log2_fold_change_list.append(log2_fold_change)
        
    log2_fold_change_list = np.array(log2_fold_change_list)
    return log2_fold_change_list
    
    
log2_fc = log2fold(uc_df,cd_df)
display(log2_fc)


array([-0.21262591, -0.06318186,  0.07587247, ..., -0.09144485,
        0.04447142,  0.22527689], shape=(10445,))